### PR TITLE
Support logging for MetricCollection with compute groups

### LINKED ISSUE #13254: Native support for Torchmetrics MetricCollection
## 🚀 Feature

Add native support in `self.log` for the `MetricCollection` object from Torchmetrics:
https://torchmetrics.readthedocs.io/en/latest/pages/overview.html#metriccollection

### Motivation

The `MetricCollection` object is used to chain together multiple metrics into a single object that can be update at once:
```python
import torchmetrics as tm
metrics = tm.MetricCollection([tm.Accuracy(), tm.Precision(), tm.Recall()])
metrics.update(preds, target)
```
however, when one want to log such an object in PL it is often done on the individual metric:
```python
self.log_dict({k:v for k,v in collection.items()})
```
However, this actually does not work with the `compute_group` feature that we [implemented in v0.8](https://devblog.pytorchlightning.ai/torchmetrics-v0-8-paper-faster-collection-and-more-metrics-4a040619eb1d) for faster computation of metrics that share the same computations.

This issue proposed that PL natively supports `MetricCollection` in the same way that `Metric` is support in `self.log`.

### Additional context

Already discussed on slack with @Borda and @justusschock 
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @akihironitta

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes #13254 

Adds native support for logging torchmetrics MetricCollection without destroying computegroups.

### Does your PR introduce any breaking changes? If yes, please list them.
None
## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @Blaizzy

### CODE REVIEW

File: tests/tests_pytorch/trainer/logging_/test_logger_connector.py [hunk 1]
```diff
@@ -506,6 +507,47 @@ def _assert_called(model, fn, stage):
...
+        def __init__(self):
+            super().__init__()
+            self.metrics = CustomMetricsCollection([Accuracy(), Precision()], compute_groups=compute_groups)
+            self.layer = torch.nn.Linear(1, 1)
+
+        def training_step(self, batch):
+
+            self.metrics(torch.rand(10, 10).softmax(-1), torch.randint(0, 10, (10,)))
+            self.log_dict(self.metrics, on_step=True, on_epoch=True)
+            return self.metrics["accuracy"].compute()
```
Comment (justusschock) [inline tests/tests_pytorch/trainer/logging_/test_logger_connector.py:531]:
```suggestion
            return batch.sum()
```

File: src/pytorch_lightning/core/module.py [hunk 2]
```diff
@@ -507,7 +512,15 @@ def log_dict(
             rank_zero_only: Whether the value will be logged only on rank 0. This will prevent synchronization which
                 would produce a deadlock as not all processes would perform this log call.
         """
-        for k, v in dictionary.items():
+
+        kwargs = {}
+
+        if isinstance(dictionary, MetricCollection):
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:518]: I'm very confused. How does logging these 2 keys do anything?

**Review by carmocca — CHANGES_REQUESTED** [review]

File: src/pytorch_lightning/core/module.py [hunk 3]
```diff
@@ -507,7 +512,15 @@ def log_dict(
...
+            kwargs["keep_base"] = False
+            if _TORCHMETRICS_GREATER_EQUAL_0_9_1 and dictionary._enable_compute_groups:
+                kwargs["copy_state"] = False
+
+        for k, v in dictionary.items(**kwargs):
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/core/module.py:523]: not for dicts (which is why in this case it is empty), but for our custom module dict (`MetricCollection`) it can.

File: src/pytorch_lightning/utilities/types.py [hunk 4]
```diff
@@ -24,14 +24,14 @@
...
 from torch import Tensor
 from torch.utils.data import DataLoader
-from torchmetrics import Metric
+from torchmetrics import Metric, MetricCollection
 from typing_extensions import Protocol, runtime_checkable
 
 from lightning_lite.utilities.types import _LRScheduler, ProcessGroup, ReduceLROnPlateau
 
 _NUMBER = Union[int, float]
 _METRIC = Union[Metric, Tensor, _NUMBER]
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/utilities/types.py:33]: I don't think so. Essentially `MetricCollection` in TM is just a dict of metrics and therefore it should fit to the definition below in PL as this is meant to only be logged with `log_dict`

[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/utilities/types.py:33]: edit: just realized I'm wrong. It should be added to neither of those.

File: tests/tests_pytorch/trainer/logging_/test_logger_connector.py [hunk 5]
```diff
@@ -506,6 +510,74 @@ def _assert_called(model, fn, stage):
...
+        def items(self, keep_base: bool = False, copy_state: bool = True):
+            if getattr(self, "_is_currently_logging", False):
+                self.wrapped_assertion_calls(keep_base, copy_state)
+
+            return super().items(keep_base=keep_base, copy_state=copy_state)
+
+    class DummyModule(LightningModule):
+        def __init__(self):
+            super().__init__()
+            if compare_version("torchmetrics", operator.ge, "0.10.0"):
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/logging_/test_logger_connector.py:533]: What's the beenfit of having this? Why not just import a metric that is present on all versions?

**Review by carmocca — APPROVED** [review]
LGTM

The override of `.items()` made me very confused. I don't think it was a good idea to reuse the API of python but have it take arguments. Anyways, that's out of the scope of this PR

File: tests/tests_pytorch/trainer/logging_/test_logger_connector.py (hunk 5 above)
Comment (justusschock) [reply tests/tests_pytorch/trainer/logging_/test_logger_connector.py:533]: We need metrics that actually can share a compute group (which are mainly the classification ones and they have all been revamped for 0.10.0).

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
Ooops, IPU had a segmentation fault. Makes total sense!

**Review by Borda — APPROVED** [review]
lgtm

**Review by SkafteNicki — APPROVED** [review]

**Comment by Borda** [comment]
@justusschock anything left? :otter:

**Comment by awaelchli** [comment]
@justusschock @Borda I fixed the mypy error but, there seems to be an issue with IPU and this implementation. There is a segmentation fault and no further information. Perhaps we can skip the test on IPU.

**Comment by carmocca** [comment]
I agree with skipping. I already had to skip several IPU tests when dropping support for PyTorch 1.9 (poptorch was using this version). Their team will have to make the effort to fix them again

**Review by tchaton — APPROVED** [review]
LGTM ! Nice addition. Can we detect a users manually called `self.log(**metric_collection.items())` and raise a warning

File: src/pytorch_lightning/utilities/imports.py [hunk 6]
```diff
@@ -23,6 +23,8 @@
 _TORCH_LESSER_EQUAL_1_10_2 = compare_version("torch", operator.le, "1.10.2")
 # duplicated from lite because HPU is patching it below
 _TORCH_GREATER_EQUAL_1_13 = compare_version("torch", operator.ge, "1.13.0")
+_TORCH_GREATER_EQUAL_2_0 = compare_version("torch", operator.ge, "2.0.0", use_base_version=True)
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/imports.py:26]: You can import this from fabric's module of the same name

```suggestion
```

**Comment by Borda** [comment]
@justusschock what is missing here? :chipmunk:

File: src/pytorch_lightning/utilities/imports.py [hunk 7]
```diff
@@ -23,6 +23,8 @@
 _TORCH_LESSER_EQUAL_1_10_2 = compare_version("torch", operator.le, "1.10.2")
 # duplicated from lite because HPU is patching it below
 _TORCH_GREATER_EQUAL_1_13 = compare_version("torch", operator.ge, "1.13.0")
+_TORCH_GREATER_EQUAL_1_14 = compare_version("torch", operator.ge, "1.14.0", use_base_version=True)
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/imports.py:26]:
```suggestion
```

File: src/pytorch_lightning/core/module.py [hunk 8]
```diff
@@ -50,7 +50,8 @@
 from pytorch_lightning.trainer.connectors.logger_connector.fx_validator import _FxValidator
 from pytorch_lightning.utilities import GradClipAlgorithmType
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
-from pytorch_lightning.utilities.imports import _TORCH_GREATER_EQUAL_1_13
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:53]: You need to keep this import. Not Fabric's

Because HPU does modify it: https://github.com/Lightning-AI/lightning/blob/a56c12c26f8f77c22984b744cf1d830a3c813cfb/src/pytorch_lightning/utilities/imports.py#L48-L49

Comment (justusschock) [reply src/pytorch_lightning/core/module.py:53]: thanks, that actually makes a lot of sense. I don't know how I could overlook this one. 🤦

File: .azure/ipu-tests.yml [hunk 9]
```diff
@@ -92,8 +92,11 @@ jobs:
       displayName: "Check poptorch installation"
 
     - bash: |
+        set -eux
         source ${{ variables.poplar_sdk }}/poplar-ubuntu*/enable.sh
         source ${{ variables.poplar_sdk }}/popart-ubuntu*/enable.sh
+        python -c "from torchmetrics import MetricCollection"
+
```
Comment (carmocca) [inline .azure/ipu-tests.yml:99]:
```suggestion
```

File: src/pytorch_lightning/core/module.py [hunk 10]
```diff
@@ -559,8 +561,19 @@ def log_dict(
                 would produce a deadlock as not all processes would perform this log call.
         """
         if self._fabric is not None:
-            return self._log_dict_through_fabric(dictionary=dictionary, logger=logger)
-        for k, v in dictionary.items():
+            return self._log_dict_through_fabric(dictionary=dictionary, logger=logger)  # type: ignore[arg-type]
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:564]:
```suggestion
            return self._log_dict_through_fabric(dictionary=dictionary, logger=logger)
```

File: src/pytorch_lightning/core/module.py [hunk 11]
```diff
@@ -559,8 +561,19 @@ def log_dict(
...
+
+        kwargs: Dict[str, bool] = {}
+
+        if isinstance(dictionary, MetricCollection):
+            kwargs["keep_base"] = False
+            if (
+                "copy_state" in inspect.signature(MetricCollection.items).parameters
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:571]: I wondered if using `RequirementCache("torchmetrics<=...")` instead of `compare_version` would have worked too.

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/core/module.py (hunk 11 above)
Comment (carmocca) [reply src/pytorch_lightning/core/module.py]: Looks like it worked.
